### PR TITLE
Add exams and statistics implementations to DG

### LINKED ISSUE #384: Mean and Median scores do not update on delete/add person

### LINKED ISSUE #386: Add all Exam related features (addExam, deleteExam, selectExam, deselectExam) to DG (Johan)

### PR DESCRIPTION
**Pull request by delishad21**
The present developer guide is lacking in documentation for our newly implemented exam features. Lets update the developer guide with explainations and diagrams detailing our exam features so that potential developers may build upon our application easily.

This PR also includes a bug fix for #384. Mean and Median scores now correctly updates on person deletetion and addition

The updated DG can be viewed **[here](https://ba44-42-60-14-21.ngrok-free.app/DeveloperGuide.html)** (lmk if the site is down)

Related to #329 
Closes #386 
Fixes #384

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -95,6 +95,21 @@ The sequence diagram below illustrates a more in-depth view of the interactions
...
+
+1. **Update using listeners embeded into UI components** - This design choice would involve embedding listeners into the UI components that would listen for changes in the `Model` (e.g. adding a listener to filteredPersons in ExamListPanel). This would allow for a more loosely coupled system, but would involve more complex implementation which could get messy as the number of listeners increase.
+2. **Update using a centralized update method** - This design choice involves having a centralized `update` method in the `MainWindow` that would call an `update` method in all other UI components after every command. This would involve a more tightly coupled system and may involve unnecessary updates, but would be easier to implement and maintain.
+
+We chose the second design choice as having a centralized update method would allow for easier maintenance, as there is a clear indicator of how UI components are updated from `MainWindow`. Adding extensions would also be more straightforward as future developers would know where to look for the update logic.
```
Comment (Zer0Legion) [inline docs/DeveloperGuide.md:107]: Great justification on why we chose the centralized method! Keeps our UI error free and foolproof.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -464,6 +479,165 @@ As the `Model` class was built prior to the implementation of this feature, we d
...
+
+Furthermore, storing the statistics allows us to maintain the code structure of our UI component, which is designed to observe and retrieve data from the `Model` component. If the statistics were to be calculated on the fly, the UI component would have to either calculate the statistics itself or request the `Model` component to calculate the statistics, which would have complicated the code structure by introducing more dependencies between the UI and Model components.
+
+##### Using `ScoreStatistics` Class
+
+The `ScoreStatistics` class was used to store the mean and median scores of the selected exam. This class was chosen as it provides a clean and structured way to store the statistics. The class also provides extensibility, as additional statistics can easily be added in the future by extending the class.
+
+<br>
+
```
Comment (Zer0Legion) [inline docs/DeveloperGuide.md:640]: Comprehensive coverage of the Exam related features.

File: src/main/java/seedu/address/model/ModelManager.java [hunk 3]
```diff
@@ -92,6 +92,7 @@ public void setAddressBookFilePath(Path addressBookFilePath) {
     @Override
     public void setAddressBook(ReadOnlyAddressBook addressBook) {
         this.addressBook.resetData(addressBook);
+        updateSelectedExamStatistics();
```
Comment (Pughal77) [inline src/main/java/seedu/address/model/ModelManager.java:95]: Nice that you have abstracted it such that all new UI components call the same function

**Review by Pughal77 — COMMENTED** [review]
Great changes to the Developer Guide, very thorough and comprehensive

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -464,6 +479,165 @@ As the `Model` class was built prior to the implementation of this feature, we d
...
+#### **Considerations for Exam Features**
+
+##### Using a selection system for exams
+
+We decided to implement a selection system for exams to complement the exam score feature. The application would only display the scores of the selected exam, making it easier for users to manage and view the scores.
+
+Our alternative design was to display the scores of all exams at once on every person. However, this alternative design would have made the UI cluttered and less user-friendly. The selection system allows users to focus on the scores of a specific exam, making it easier to view and manage the scores.
```
Comment (jayllo-c) [inline docs/DeveloperGuide.md:586]: Good elaboration on the alternatives we could have gone for and why we chose to follow our current implementation.

File: docs/DeveloperGuide.md
Comment (jayllo-c) [inline docs/DeveloperGuide.md:1]: Exam features have been explained thoroughly and logically

**Review by jayllo-c — COMMENTED** [review]
Good job on the DG changes.

**Pull request merged**